File: src/core/format.ts
const MINUTE = 60_000
const HOUR = 60 * MINUTE
const DAY = 24 * HOUR

/**
 * Elapsed time as the coarsest unit that fits: `10d`, `3h`, `30m`, or `<1m`
 * below a minute, since neither caller has room for `0m`.
 */
function elapsedSince(iso: string, now: string): string {
  const elapsed = Date.parse(now) - Date.parse(iso)
  if (elapsed < MINUTE) return '<1m'
  if (elapsed < HOUR) return `${Math.floor(elapsed / MINUTE)}m`
  if (elapsed < DAY) return `${Math.floor(elapsed / HOUR)}h`
  return `${Math.floor(elapsed / DAY)}d`
}

export function formatAge(iso: string, now: string): string {
  const elapsed = Date.parse(now) - Date.parse(iso)
  if (elapsed < MINUTE) return 'just now'
  return `${elapsedSince(iso, now)} ago`
}

/**
 * How long a pull request has been waiting on the user — `waiting 10d`, which
 * says what `10d ago` cannot: that the ball has been in their court that whole
 * time, not merely that somebody touched the thread then.
 */
export function formatWaiting(iso: string, now: string): string {
  return `waiting ${elapsedSince(iso, now)}`
}

/**
 * How long until `iso`, as words for a sentence like "try again in
 * ${formatWait(...)}" — the mirror image of `formatAge`, needed because a
 * rate-limit reset is the one place Pullover talks about a moment still to
 * come rather than one already past. `formatAge(now, iso)` can't be reused
 * for this: swapping its arguments produces the right *magnitude* but the
 * wrong words ("12m ago" for a reset that hasn't happened yet), so the
 * bucketing is duplicated here rather than the suffix being patched in from
 * outside.
 *
 * Minutes round up, not down like `formatAge`'s elapsed time, so the wait
 * named here is never shorter than the real one — a user who retries right
 * when the countdown hits zero must not still get refused.
 */
export function formatWait(iso: string, now: string): string {
  const remaining = Date.parse(iso) - Date.parse(now)
  if (remaining <= MINUTE) return 'a minute'
  if (remaining < HOUR) return `${Math.ceil(remaining / MINUTE)} minutes`
  const hours = Math.ceil(remaining / HOUR)
  return hours === 1 ? '1 hour' : `${hours} hours`
}

/**
 * Just the repository, dropping the owner `nameWithOwner` carries.
 *
 * Only for display: `pr.repository` stays the full name everywhere else,
 * because that is what groups a stack, what the watch list matches against
 * and what the settings picker lists — an owner-less name would collide
 * between two organisations.
 */
export function repositoryName(fullName: string): string {
  const slash = fullName.lastIndexOf('/')
  return slash === -1 ? fullName : fullName.slice(slash + 1)
}

File: Sources/PulloverCore/Format.swift
import Foundation

private let minute: TimeInterval = 60
private let hour: TimeInterval = 60 * minute
private let day: TimeInterval = 24 * hour

/// Elapsed time as the coarsest unit that fits: `10d`, `3h`, `30m`, or `<1m`
/// below a minute, since neither caller has room for `0m`.
private func elapsed(since date: Date, now: Date) -> String {
    let elapsed = now.timeIntervalSince(date)
    if elapsed < minute { return "<1m" }
    if elapsed < hour { return "\(Int(elapsed / minute))m" }
    if elapsed < day { return "\(Int(elapsed / hour))h" }
    return "\(Int(elapsed / day))d"
}

public func formatAge(_ date: Date, now: Date) -> String {
    if now.timeIntervalSince(date) < minute { return "just now" }
    return "\(elapsed(since: date, now: now)) ago"
}

/// How long a pull request has been waiting on the user — `waiting 10d`, which
/// says what `10d ago` cannot: that the ball has been in their court all that
/// time, not that somebody touched the thread then.
public func formatWaiting(_ date: Date, now: Date) -> String {
    "waiting \(elapsed(since: date, now: now))"
}

/// How long until `date`, as words for "try again in …". Minutes round up, so
/// the wait named is never shorter than the real one — a user who retries
/// when the countdown hits zero must not still get refused.
public func formatWait(until date: Date, now: Date) -> String {
    let remaining = date.timeIntervalSince(now)
    if remaining <= minute { return "a minute" }
    if remaining < hour { return "\(Int((remaining / minute).rounded(.up))) minutes" }
    let hours = Int((remaining / hour).rounded(.up))
    return hours == 1 ? "1 hour" : "\(hours) hours"
}

/// Just the repository, dropping the owner. Only for display: everywhere else
/// the full name is what groups a stack and what the watch list matches.
public func repositoryName(_ fullName: String) -> String {
    guard let slash = fullName.lastIndex(of: "/") else { return fullName }
    return String(fullName[fullName.index(after: slash)...])
}

/// Two letters: one is ambiguous at a glance across a list of teammates.
public func initials(of login: String) -> String {
    String(login.prefix(2)).uppercased()
}
